### PR TITLE
167 basic tracer

### PR DESCRIPTION
**Pull request by CodiePP**
description
-----------

- [x] include new `Tracer`, prepare code (parameterisation)

checklist
---------

- [x] compiles (`cabal new-clean; cabal new-build`)
- [x] tests run successfully (`cabal new-test`)
- [x] documentation added and created (`cd docs; nix-shell --run make`)
- [x] link to an issue
- [x] link to an epic
- [x] add estimate points
- [x] add milestone (the same as the linked issue)

### CODE REVIEW

File: src/Cardano/BM/Tracer/Class.lhs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+\end{code}
+
+The notion of a `Tracer` is an action that can be used to observe
+information of interest during evaluation. `Tracer`s can capture (and
+annotate) such observations with additional information from their
+execution context.
+
+\begin{code}
+newtype Tracer m s = Tracer { tracing :: Op (m ()) s }
```
Comment (njd42) [inline src/Cardano/BM/Tracer/Class.lhs]: We are using `Op` here, which is a slight complexity over the version we discussed last week. Realise that @avieth did this in past; could see that instances of `Op` might make certain transformations (e.g. collecting sequences of traced items into a monoid) - just want to make sure that we clear why the additional complexity.

File: src/Cardano/BM/Tracer/Class.lhs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+The simplest tracer - one that does not generate output
+
+\begin{code}
+nullTracer :: (Applicative m) => Tracer m a
+nullTracer = Tracer $ Op $ \_ -> pure ()
+
+\end{code}
+
+
+Accepts a |Trace| and some payload |s|. First it gets the contravariant
```
Comment (njd42) [inline src/Cardano/BM/Tracer/Class.lhs]: |Trace|  needs to be |Tracer|

File: src/Cardano/BM/Tracer/Class.lhs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+from the |Trace| as type "|Op (m ()) s|" and, after "|getOp :: b -> a|" which
+translates to "|s -> m ()|".
+\begin{code}
+traceWith :: Tracer m s -> s -> m ()
```
Comment (njd42) [inline src/Cardano/BM/Tracer/Class.lhs]: maybe the names should be `tracingWith` - the idea here being that we are distinguishing this notion of "Trace" from other notions of "Trace" by appealing to it being in the present tense - ie you are _tracing with a tracer_

File: src/Cardano/BM/Tracer/Transformers.lhs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+Trace creation time.
+
+\begin{code}
+condTracing :: (Monad m) => Bool -> Tracer m a -> Tracer m a
+condTracing active tr
+  | active    = tr
+  | otherwise = nullTracer
+\end{code}
+
+The Tracer transformer with dynamic control of tracing.
```
Comment (njd42) [inline src/Cardano/BM/Tracer/Transformers.lhs]: The tracer transformer that can exercise dynamic control over tracing, the dynamic decision being made using the context accessible in the monadic context.

File: src/Cardano/BM/Tracer/Transformers.lhs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+\begin{code}
+condTracingM :: (Monad m) => m Bool -> Tracer m a -> Tracer m a
```
Comment (njd42) [inline src/Cardano/BM/Tracer/Transformers.lhs]: I think a better (and more useful) example would be one with the type signature:

`condTracingM :: (Monad m) => m (a -> Bool) -> Tracer m a -> Tracer m a`

which can be trivially reduced to this case using `const`

File: src/Cardano/BM/Tracer/Transformers.lhs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+showTracing :: (Show a) => Tracer m String -> Tracer m a
+showTracing = contramap show
+\end{code}
+
+
+The Tracer transformer that allows for on/off control of tracing at
+Trace creation time.
+
+\begin{code}
+condTracing :: (Monad m) => Bool -> Tracer m a -> Tracer m a
```
Comment (njd42) [inline src/Cardano/BM/Tracer/Transformers.lhs]: see below, maybe this type signature should be 

`condTracing :: (Monad m) => (a -> Bool) -> Tracer m a -> Tracer m a`

**Comment by CodiePP** [comment]
some changes have been moved to issue #168 
the `Tracer` interface that is introduced here will be moved to its own repo.
but first, we have to merge this PR and then we can go on.

**Comment by njd42** [comment]
lgtm

**Pull request merged**